pyomo/contrib/pynumero/interfaces/pyomo_nlp.py

```diff
@@ -298,34 +299,24 @@ def get_inequality_constraint_indices(self, constraints):
     # overloaded from NLP
     def get_obj_scaling(self):
         obj = self.get_pyomo_objective()
-        scaling_suffix = self._pyomo_model.component('scaling_factor')
-        if scaling_suffix and scaling_suffix.ctype is pyo.Suffix:
-            if obj in scaling_suffix:
-                return scaling_suffix[obj]
-            return 1.0
-        return None
+        val = SuffixFinder('scaling_factor', 1.0).find(obj)
+        return val
 
     # overloaded from NLP
     def get_primals_scaling(self):
-        scaling_suffix = self._pyomo_model.component('scaling_factor')
-        if scaling_suffix and scaling_suffix.ctype is pyo.Suffix:
-            primals_scaling = np.ones(self.n_primals())
-            for i, v in enumerate(self.get_pyomo_variables()):
-                if v in scaling_suffix:
-                    primals_scaling[i] = scaling_suffix[v]
-            return primals_scaling
-        return None
+        scaling_suffix_finder = SuffixFinder('scaling_factor', 1.0)
+        primals_scaling = np.ones(self.n_primals())
+        for i, v in enumerate(self.get_pyomo_variables()):
+            primals_scaling[i] = scaling_suffix_finder.find(v)
+        return primals_scaling
```

Comment:
Returning ones instead of `None` here when no scaling factors have been defined is the backward incompatibility I'm concerned about. Maybe something like this would be better?
```python
scaling_suffix_finder = SuffixFinder("scaling_factor")
scaling_factors = [scaling_suffix_finder.find(v) for v in self.get_pyomo_variables()]
has_scaling_factors = any(sf is not None for sf in scaling_factors)
if has_scaling_factors:
    return [sf if sf is not None else 1.0 for sf in scaling_factors]
else:
    return None
```
It's not exactly the same as having any non-empty `scaling_factor` suffix used to be enough to avoid returning `None`. Now we need to actually contain one of the primal variables. Maybe we could make the `SuffixFinder._get_suffix_list` method public, and return `None` if all these suffixes are empty?

Reply:
I implemented something equivalent originally (see https://github.com/Pyomo/pyomo/pull/3295/commits/23d316e857b35e81d990114a918f215e5e3a556e) but this will break the existing "hacks" in the implicit function solver: https://github.com/Pyomo/pyomo/blob/ae354aeb7d62b62b049143be84ac6a7437bbe44d/pyomo/contrib/pynumero/algorithms/solvers/implicit_functions.py#L406-L410

To support both cases this function could an optional argument `always_return_numeric=False` or something similar to determine which behavior is desired. Maybe you have a better idea about how to de-conflict?

Reply:
Yes, if we do this, I'll have to explicitly set scaling factors to 1 in the implicit function solver, which wouldn't be the worst thing ever. I think explicitly checking `SuffixFinder._get_suffix_list` for at least one non-empty suffix would get around this as well.

Reply:
In 9ca3928 I have opted just to replicate the original check. This seems the safest and doesn't involve accessing private methods or modifying other parts of Pyomo.